### PR TITLE
Fix: Streaming server memory leak in HTTP EventSource cleanup

### PR DESCRIPTION
**Pull request by ThisIsMissEm**
This was discovered in #24949, but never got the attention it needed. This pulls the fix out into a separate PR: https://github.com/mastodon/mastodon/pull/24949/commits/74b22f2af7eacf3eb46516f943fd135b71736d41

Essentially when EventSource streams ended, they failed to actually remove the listener from subscription callbacks/listeners invoked by Redis receiving a new event.

The change to `attachCloseHandler` was just to ensure type safety (it was giving a type assert when trying to verify I'd correctly handled the unsubscribe correctly everywhere):
```diff
-    if (attachCloseHandler) {
+    if (typeof attachCloseHandler === 'function') {
```

### CODE REVIEW

File: streaming/index.js [hunk 1]
```diff
@@ -222,9 +222,15 @@ const startServer = async () => {
     callbacks.forEach(callback => callback(json));
   };
 
+  /**
+   * @callback SubscriptionListener
+   * @param {ReturnType<parseJSON>} json of the message
+   * @returns voids
```
Comment (ClearlyClaire) [inline streaming/index.js:228]: `voids`?

Comment (ThisIsMissEm) [reply streaming/index.js:228]:
```suggestion
   * @returns void
```

**Review by ClearlyClaire — APPROVED** [review]
Looks good but I'd really like another set of eyes here, especially if this is going to be included in the upcoming releases.

**Review by 868144 — COMMENTED** [review]
> good

**Pull request merged**